File: app/src/main/java/github/yaa110/memento/fragment/template/RecyclerFragment.java
package github.yaa110.memento.fragment.template;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Locale;

import github.yaa110.memento.R;
import github.yaa110.memento.adapter.template.ModelAdapter;
import github.yaa110.memento.db.Controller;
import github.yaa110.memento.db.OpenHelper;
import github.yaa110.memento.inner.Animator;
import github.yaa110.memento.model.Category;
import github.yaa110.memento.model.DatabaseModel;
import github.yaa110.memento.model.Note;

abstract public class RecyclerFragment<T extends DatabaseModel, A extends ModelAdapter> extends Fragment {
    public View fab;
    public Toolbar selectionToolbar;
    public boolean selectionState = false;
    public ArrayList<T> items;
    public ArrayList<T> selected = new ArrayList<>();
    public Callbacks activity;
    public long categoryId = DatabaseModel.NEW_MODEL_ID;
    public String categoryTitle;
    public int categoryTheme;
    public int categoryPosition = 0;
    private RecyclerView recyclerView;
    private View empty;
    private TextView selectionCounter;
    private A adapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getLayout(), container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        fab = view.findViewById(R.id.fab);
        recyclerView = view.findViewById(R.id.recyclerView);
        empty = view.findViewById(R.id.empty);
        selectionToolbar = getActivity().findViewById(R.id.selection_toolbar);
        selectionCounter = selectionToolbar.findViewById(R.id.selection_counter);

        init(view);

        selectionToolbar.findViewById(R.id.selection_back).setOnClickListener(view3 -> toggleSelection(false));

        selectionToolbar.findViewById(R.id.selection_delete).setOnClickListener(view4 -> {
            final ArrayList<T> undos = new ArrayList<>(selected);
            toggleSelection(false);

            new Thread() {
                @Override
                public void run() {
                    final int length = undos.size();
                    String[] ids = new String[length];
                    final int[] sortablePosition = new int[length];

                    for (int i = 0; i < length; i++) {
                        T item = undos.get(i);
                        ids[i] = String.format(Locale.US, "%d", item.id);
                        int position = items.indexOf(item);
                        item.position = position;
                        sortablePosition[i] = position;
                    }

                    Controller.instance.deleteNotes(ids, categoryId);

                    Arrays.sort(sortablePosition);

                    getActivity().runOnUiThread(() -> {
                        for (int i = length - 1; i >= 0; i--) {
                            items.remove(sortablePosition[i]);
                            adapter.notifyItemRemoved(sortablePosition[i]);
                        }

                        toggleEmpty();

                        StringBuilder message = new StringBuilder();
                        message.append(length).append(" ").append(getItemName());
                        if (length > 1) message.append("s were deleted");
                        else message.append(" was deleted.");

                        Snackbar.make(fab != null ? fab : selectionToolbar, message.toString(), 7000)
                                .setAction(R.string.undo, view1 -> new Thread() {
                                    @Override
                                    public void run() {
                                        Controller.instance.undoDeletion();
                                        if (categoryId != DatabaseModel.NEW_MODEL_ID) {
                                            Controller.instance.addCategoryCounter(categoryId, length);
                                        }

                                        Collections.sort(undos, (t1, t2) -> {
                                            if (t1.position < t2.position) return -1;
                                            if (t1.position == t2.position) return 0;
                                            return 1;
                                        });

                                        getActivity().runOnUiThread(() -> {
                                            for (int i = 0; i < length; i++) {
                                                T item = undos.get(i);
                                                addItem(item, item.position);
                                            }
                                        });
                                        interrupt();
                                    }
                                }.start())
                                .show();
                    });

                    interrupt();
                }
            }.start();
        });

        if (fab != null) {
            fab.setOnClickListener(view2 -> onClickFab());
        }

        Intent data = getActivity().getIntent();
        if (data != null) {
            // Get the parent data
            categoryId = data.getLongExtra(OpenHelper.COLUMN_ID, DatabaseModel.NEW_MODEL_ID);
            categoryTitle = data.getStringExtra(OpenHelper.COLUMN_TITLE);
            categoryTheme = data.getIntExtra(OpenHelper.COLUMN_THEME, Category.THEME_GREEN);
            categoryPosition = data.getIntExtra("position", 0);

            if (categoryTitle != null) {
                ((TextView) getActivity().findViewById(R.id.title)).setText(categoryTitle);
            }
        }

        loadItems();
    }

    public void onChangeCounter(int count) {
        selectionCounter.setText(String.format(Locale.US, "%d", count));
    }

    public void toggleSelection(boolean state) {
        selectionState = state;
        activity.onChangeSelection(state);
        if (state) {
            Animator.create(getContext())
                    .on(selectionToolbar)
                    .setStartVisibility(View.VISIBLE)
                    .animate(R.anim.fade_in);
        } else {
            Animator.create(getContext())
                    .on(selectionToolbar)
                    .setEndVisibility(View.GONE)
                    .animate(R.anim.fade_out);

            deselectAll();
        }
    }

    private void deselectAll() {
        while (!selected.isEmpty()) {
            adapter.notifyItemChanged(items.indexOf(selected.remove(0)));
        }
    }

    public void loadItems() {
        new Thread() {
            @SuppressWarnings("unchecked")
            @Override
            public void run() {
                try {
                    if (categoryId == DatabaseModel.NEW_MODEL_ID) {
                        // Get all categories
                        items = (ArrayList<T>) Category.all();
                    } else {
                        // Get notes of the category by categoryId
                        items = (ArrayList<T>) Note.all(categoryId);
                    }

                    adapter = getAdapterClass().getDeclaredConstructor(
                            ArrayList.class,
                            ArrayList.class,
                            ModelAdapter.ClickListener.class
                    ).newInstance(items, selected, getListener());

                    getActivity().runOnUiThread(() -> {
                        toggleEmpty();

                        recyclerView.setAdapter(adapter);
                        recyclerView.setLayoutManager(new LinearLayoutManager(
                                getContext(),
                                LinearLayoutManager.VERTICAL,
                                false
                        ));
                    });
                } catch (Exception ignored) {
                } finally {
                    interrupt();
                }
            }
        }.start();
    }

    private void toggleEmpty() {
        if (items.isEmpty()) {
            empty.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
        } else {
            empty.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }
    }

    public void refreshItem(int position) {
        adapter.notifyItemChanged(position);
    }

    public T deleteItem(int position) {
        T item = items.remove(position);
        adapter.notifyItemRemoved(position);
        toggleEmpty();
        return item;
    }

    public void addItem(T item, int position) {
        if (items.isEmpty()) {
            empty.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }

        items.add(position, item);
        adapter.notifyItemInserted(position);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.activity = (Callbacks) context;
    }

    public void init(View view) {
    }

    public abstract void onClickFab();

    public abstract int getLayout();

    public abstract String getItemName();

    public abstract Class<A> getAdapterClass();

    public abstract ModelAdapter.ClickListener getListener();

    public interface Callbacks {
        void onChangeSelection(boolean state);

        void toggleOneSelection(boolean state);
    }
}

File: app/src/main/java/github/yaa110/memento/fragment/template/RecyclerFragment.kt
package github.yaa110.memento.fragment.template

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import github.yaa110.memento.R
import github.yaa110.memento.adapter.template.ModelAdapter
import github.yaa110.memento.db.Controller
import github.yaa110.memento.db.OpenHelper
import github.yaa110.memento.inner.Animator
import github.yaa110.memento.model.Category
import github.yaa110.memento.model.DatabaseModel
import github.yaa110.memento.model.Note
import java.util.Arrays
import java.util.Collections
import java.util.Locale

abstract class RecyclerFragment<T : DatabaseModel?, A : ModelAdapter<*, *>?> : Fragment() {
    lateinit var fab: View
    lateinit var selectionToolbar: Toolbar
    var selectionState: Boolean = false
    var items: ArrayList<T>? = null
    var selected: ArrayList<T> = ArrayList()
    var activity: Callbacks? = null
    var categoryId: Long = DatabaseModel.NEW_MODEL_ID
    var categoryTitle: String? = null
    var categoryTheme: Int = 0
    var categoryPosition: Int = 0
    private lateinit var recyclerView: RecyclerView
    private lateinit var empty: View
    private lateinit var selectionCounter: TextView
    private var adapter: A? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        fab = view.findViewById(R.id.fab)
        recyclerView = view.findViewById(R.id.recyclerView)
        empty = view.findViewById(R.id.empty)
        selectionToolbar = requireActivity().findViewById(R.id.selection_toolbar)
        selectionCounter = selectionToolbar.findViewById(R.id.selection_counter)

        init(view)

        selectionToolbar.findViewById<View>(R.id.selection_back)
            .setOnClickListener { view3: View? ->
                toggleSelection(
                    false
                )
            }

        selectionToolbar.findViewById<View>(R.id.selection_delete)
            .setOnClickListener { view4: View? ->
                val undos = ArrayList(selected)
                toggleSelection(false)
                object : Thread() {
                    override fun run() {
                        val length = undos.size
                        val ids = arrayOfNulls<String>(length)
                        val sortablePosition = IntArray(length)

                        for (i in 0..<length) {
                            val item = undos[i]
                            ids[i] = String.format(Locale.US, "%d", item!!.id)
                            val position = items!!.indexOf(item)
                            item.position = position
                            sortablePosition[i] = position
                        }

                        Controller.instance!!.deleteNotes(
                            ids,
                            categoryId
                        )

                        Arrays.sort(sortablePosition)

                        requireActivity().runOnUiThread {
                            for (i in length - 1 downTo 0) {
                                items!!.removeAt(sortablePosition[i])
                                adapter!!.notifyItemRemoved(sortablePosition[i])
                            }
                            toggleEmpty()

                            val message = StringBuilder()
                            message.append(length).append(" ").append(this@RecyclerFragment.itemName)
                            if (length > 1) message.append("s were deleted")
                            else message.append(" was deleted.")
                            Snackbar.make(
                                (if (fab != null) fab else selectionToolbar),
                                message.toString(),
                                7000
                            )
                                .setAction(R.string.undo) { view1: View? ->
                                    object : Thread() {
                                        override fun run() {
                                            Controller.instance!!.undoDeletion()
                                            if (categoryId != DatabaseModel.NEW_MODEL_ID) {
                                                Controller.instance!!.addCategoryCounter(
                                                    categoryId,
                                                    length
                                                )
                                            }

                                            Collections.sort(undos) { t1: T, t2: T ->
                                                if (t1!!.position < t2!!.position) return@sort -1
                                                if (t1.position == t2.position) return@sort 0
                                                1
                                            }

                                            requireActivity().runOnUiThread {
                                                for (i in 0..<length) {
                                                    val item = undos[i]
                                                    addItem(item, item!!.position)
                                                }
                                            }
                                            interrupt()
                                        }
                                    }.start()
                                }
                                .show()
                        }

                        interrupt()
                    }
                }.start()
            }

        fab.setOnClickListener { view2: View? -> onClickFab() }

        val data = requireActivity().intent
        if (data != null) {
            // Get the parent data
            categoryId = data.getLongExtra(
                OpenHelper.COLUMN_ID,
                DatabaseModel.NEW_MODEL_ID
            )
            categoryTitle = data.getStringExtra(OpenHelper.COLUMN_TITLE)
            categoryTheme =
                data.getIntExtra(OpenHelper.COLUMN_THEME, Category.THEME_GREEN)
            categoryPosition = data.getIntExtra("position", 0)

            if (categoryTitle != null) {
                (requireActivity().findViewById<View>(R.id.title) as TextView).text =
                    categoryTitle
            }
        }

        loadItems()
    }

    fun onChangeCounter(count: Int) {
        selectionCounter.text = String.format(Locale.US, "%d", count)
    }

    fun toggleSelection(state: Boolean) {
        selectionState = state
        activity!!.onChangeSelection(state)
        if (state) {
            Animator.create(context)
                .on<Toolbar?>(selectionToolbar)
                .setStartVisibility(View.VISIBLE)
                .animate(R.anim.fade_in)
        } else {
            Animator.create(context)
                .on<Toolbar?>(selectionToolbar)
                .setEndVisibility(View.GONE)
                .animate(R.anim.fade_out)

            deselectAll()
        }
    }

    private fun deselectAll() {
        while (selected.isNotEmpty()) {
            adapter!!.notifyItemChanged(items!!.indexOf(selected.removeAt(0)))
        }
    }

    fun loadItems() {
        object : Thread() {
            override fun run() {
                try {
                    items = if (categoryId == DatabaseModel.NEW_MODEL_ID) {
                        // Get all categories
                        Category.all() as ArrayList<T>
                    } else {
                        // Get notes of the category by categoryId
                        Note.all(categoryId) as ArrayList<T>
                    }

                    adapter = this@RecyclerFragment.adapterClass.getDeclaredConstructor(
                        ArrayList::class.java,
                        ArrayList::class.java,
                        ModelAdapter.ClickListener::class.java
                    ).newInstance(items, selected, this@RecyclerFragment.listener)

                    requireActivity().runOnUiThread {
                        toggleEmpty()
                        recyclerView.adapter = adapter
                        recyclerView.layoutManager = LinearLayoutManager(
                            context,
                            LinearLayoutManager.VERTICAL,
                            false
                        )
                    }
                } catch (ignored: Exception) {
                } finally {
                    interrupt()
                }
            }
        }.start()
    }

    private fun toggleEmpty() {
        if (items!!.isEmpty()) {
            empty.visibility = View.VISIBLE
            recyclerView.visibility = View.INVISIBLE
        } else {
            empty.visibility = View.GONE
            recyclerView.visibility = View.VISIBLE
        }
    }

    fun refreshItem(position: Int) {
        adapter!!.notifyItemChanged(position)
    }

    fun deleteItem(position: Int): T {
        val item = items!!.removeAt(position)
        adapter!!.notifyItemRemoved(position)
        toggleEmpty()
        return item
    }

    fun addItem(item: T, position: Int) {
        if (items!!.isEmpty()) {
            empty.visibility = View.GONE
            recyclerView.visibility = View.VISIBLE
        }

        items!!.add(position, item)
        adapter!!.notifyItemInserted(position)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.activity = context as Callbacks
    }

    open fun init(view: View) {
    }

    abstract fun onClickFab()

    abstract val layout: Int

    abstract val itemName: String

    abstract val adapterClass: Class<A>

    abstract val listener: ModelAdapter.ClickListener<*>?

    interface Callbacks {
        fun onChangeSelection(state: Boolean)

        fun toggleOneSelection(state: Boolean)
    }
}
